// steam-provider/dist/services/steam/models/steam-entry.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/**
 * Class for the info for each search result
 */
class SteamSearchEntry {
    constructor($id, $name, $url, $price, $priceWithDiscount, $score) {
        this.id = $id;
        this.name = $name;
        this.url = $url;
        this.price = $price;
        this.priceWithDiscount = $priceWithDiscount;
        this.score = $score;
    }
    get $id() {
        return this.id;
    }
    get $name() {
        return this.name;
    }
    get $url() {
        return this.url;
    }
    get $price() {
        return this.price;
    }
    get $priceWithDiscount() {
        return this.priceWithDiscount;
    }
    get $score() {
        return this.score;
    }
}
exports.SteamSearchEntry = SteamSearchEntry;
/**
 * Class for the details  of each game
 */
class SteamDetailEntry {
    constructor($id, $name, $genres, $priceData, $otherData) {
        this.id = $id;
        this.name = $name;
        this.genres = $genres;
        this.priceData = $priceData;
        this.otherData = $otherData;
    }
    get $id() {
        return this.id;
    }
    get $name() {
        return this.name;
    }
    get $genres() {
        return this.genres;
    }
    get $priceData() {
        return this.priceData;
    }
    get $otherData() {
        return this.otherData;
    }
}
exports.SteamDetailEntry = SteamDetailEntry;
/**
 * Other info inside SteamDetailEntry
 */
class OtherInfo {
    constructor($imageUrl, $platforms, $metacriticScore, $features, $developer, $publisher) {
        this.imageUrl = $imageUrl;
        this.platforms = $platforms;
        this.metacriticScore = $metacriticScore;
        this.features = $features;
        this.developer = $developer;
        this.publisher = $publisher;
    }
    get $imageUrl() {
        return this.imageUrl;
    }
    get $platforms() {
        return this.platforms;
    }
    get $metacriticScore() {
        return this.metacriticScore;
    }
    get $features() {
        return this.features;
    }
    get $developer() {
        return this.developer;
    }
    get $publisher() {
        return this.publisher;
    }
}
exports.OtherInfo = OtherInfo;
/**
 * Pricing info inside SteamDetailEntry
 */
class PriceInfo {
    constructor($initialPrice, $finalPrice, $discountPercent) {
        this.initialPrice = $initialPrice;
        this.finalPrice = $finalPrice;
        this.discountPercent = $discountPercent;
    }
    get $initialPrice() {
        return this.initialPrice;
    }
    get $finalPrice() {
        return this.finalPrice;
    }
    get $discountPercent() {
        return this.discountPercent;
    }
}
exports.PriceInfo = PriceInfo;
